File: bot/domain/entities/library.py
from __future__ import annotations
from typing import List, Optional

from .playlist import Playlist
from ..repositories.playlist_repository import PlaylistRepository
from ..valueobjects.source_type import SourceType
from ...pkg.logger import logger


class LibraryManager:
    """Manages the music library with persistent storage"""

    def __init__(self, base_path: str = "playlist"):
        self._repository = PlaylistRepository(base_path)
        self._cache: dict[str, Playlist] = {}

    def create_playlist(self, name: str) -> bool:
        """Create a new empty playlist"""
        if self._repository.exists(name):
            return False

        playlist = Playlist(name)
        success = self._repository.save(playlist)
        if success:
            self._cache[name] = playlist
        return success

    def get_playlist(self, name: str) -> Optional[Playlist]:
        """Retrieve a playlist by name (with caching)"""
        if name in self._cache:
            return self._cache[name]

        playlist = self._repository.load(name)
        if playlist:
            self._cache[name] = playlist
        return playlist

    def save_playlist(self, playlist: Playlist) -> bool:
        """Save playlist to storage"""
        success = self._repository.save(playlist)
        if success:
            self._cache[playlist.name] = playlist
        return success

    def delete_playlist(self, name: str) -> bool:
        """Delete a playlist"""
        success = self._repository.delete(name)
        if success and name in self._cache:
            del self._cache[name]
        return success

    def list_playlists(self) -> List[str]:
        """List all playlist names"""
        return self._repository.list_all()

    def add_to_playlist(
        self,
        playlist_name: str,
        original_input: str,
        source_type: SourceType,
        title: Optional[str] = None,
    ) -> tuple[bool, bool]:
        """Add song to playlist

        Returns:
            tuple[bool, bool]: (success, is_duplicate)
            - (True, False): Added successfully
            - (True, True): Duplicate found, not added
            - (False, False): Playlist not found or save failed
        """
        playlist = self.get_playlist(playlist_name)
        if not playlist:
            return False, False

        # Check for duplicate before adding
        if playlist.has_entry(original_input):
            logger.info(
                f"Duplicate '{original_input}' skipped in playlist '{playlist_name}'"
            )
            return True, True  # Success = True (no error), Duplicate = True

        # Add entry (returns True if successful)
        if playlist.add_entry(original_input, source_type, title):
            if self.save_playlist(playlist):
                return True, False  # Added successfully
            else:
                # Revert the add if save failed
                playlist._entries.pop()
                return False, False

        return False, False

    def remove_from_playlist(self, playlist_name: str, index: int) -> bool:
        """Remove song from playlist by index"""
        playlist = self.get_playlist(playlist_name)
        if not playlist:
            logger.error(f"Playlist '{playlist_name}' not found")
            return False

        if playlist.remove_entry(index):
            return self.save_playlist(playlist)
        logger.error(f"Failed to remove entry at index {index} from '{playlist_name}'")
        return False

    def clear_playlist(self, playlist_name: str) -> bool:
        """Clear all songs from playlist"""
        playlist = self.get_playlist(playlist_name)
        if not playlist:
            return False

        playlist.clear()
        return self.save_playlist(playlist)

    def exists(self, playlist_name: str) -> bool:
        """Check if a playlist exists"""
        return self._repository.exists(playlist_name)

File: internal/domain/entities/library.go
package entities

import (
	"sync"

	"github.com/vuongmanhnghia/discord-music-bot/internal/domain/valueobjects"
)

// PlaylistRepository interface for persistence
type PlaylistRepository interface {
	Save(playlist *Playlist) error
	Load(playlistName string) (*Playlist, error)
	Delete(playlistName string) error
	Exists(playlistName string) bool
	ListAll() ([]string, error)
}

// Library manages the music library with persistent storage
type Library struct {
	repository PlaylistRepository
	cache      map[string]*Playlist
	mu         sync.RWMutex
}

// NewLibrary creates a new library instance
func NewLibrary(repository PlaylistRepository) *Library {
	return &Library{
		repository: repository,
		cache:      make(map[string]*Playlist),
	}
}

// CreatePlaylist creates a new empty playlist
func (l *Library) CreatePlaylist(name string) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.repository.Exists(name) {
		return ErrPlaylistExists
	}

	playlist := NewPlaylist(name)
	if err := l.repository.Save(playlist); err != nil {
		return err
	}

	l.cache[name] = playlist
	return nil
}

// GetPlaylist retrieves a playlist by name (with caching)
func (l *Library) GetPlaylist(name string) (*Playlist, error) {
	l.mu.RLock()
	if cached, ok := l.cache[name]; ok {
		l.mu.RUnlock()
		return cached, nil
	}
	l.mu.RUnlock()

	l.mu.Lock()
	defer l.mu.Unlock()

	// Double-check after acquiring write lock
	if cached, ok := l.cache[name]; ok {
		return cached, nil
	}

	playlist, err := l.repository.Load(name)
	if err != nil {
		return nil, err
	}

	if playlist == nil {
		return nil, ErrPlaylistNotFound
	}

	l.cache[name] = playlist
	return playlist, nil
}

// SavePlaylist saves a playlist to storage
func (l *Library) SavePlaylist(playlist *Playlist) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if err := l.repository.Save(playlist); err != nil {
		return err
	}

	l.cache[playlist.Name] = playlist
	return nil
}

// DeletePlaylist deletes a playlist
func (l *Library) DeletePlaylist(name string) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if err := l.repository.Delete(name); err != nil {
		return err
	}

	delete(l.cache, name)
	return nil
}

// ListPlaylists returns all playlist names
func (l *Library) ListPlaylists() ([]string, error) {
	return l.repository.ListAll()
}

// AddToPlaylist adds a song to a playlist
func (l *Library) AddToPlaylist(playlistName, originalInput string, sourceType valueobjects.SourceType, title string) (bool, error) {
	playlist, err := l.GetPlaylist(playlistName)
	if err != nil {
		return false, err
	}

	// Check for duplicate
	if playlist.HasEntry(originalInput) {
		return true, nil // Not an error, just a duplicate
	}

	playlist.AddEntry(originalInput, sourceType, title)

	if err := l.SavePlaylist(playlist); err != nil {
		return false, err
	}

	return false, nil // Not a duplicate, added successfully
}

// RemoveFromPlaylist removes a song from a playlist
func (l *Library) RemoveFromPlaylist(playlistName, originalInput string) error {
	playlist, err := l.GetPlaylist(playlistName)
	if err != nil {
		return err
	}

	if !playlist.RemoveEntry(originalInput) {
		return ErrSongNotInPlaylist
	}

	return l.SavePlaylist(playlist)
}

// Custom errors
var (
	ErrPlaylistExists    = &LibraryError{Message: "playlist already exists"}
	ErrPlaylistNotFound  = &LibraryError{Message: "playlist not found"}
	ErrSongNotInPlaylist = &LibraryError{Message: "song not in playlist"}
)

// LibraryError represents a library-specific error
type LibraryError struct {
	Message string
}

func (e *LibraryError) Error() string {
	return e.Message
}
